projects/chaos/test/test_power_path.c

```diff
@@ -62,34 +80,82 @@ static PowerPathCfg s_ppc = { .enable_pin = { .port = 0, .pin = 0 },
                                         .timer_id = SOFT_TIMER_INVALID_TIMER,
                                         .monitoring_active = false } };
 
+static CANStorage s_can_storage;
+static CANRxHandler s_rx_handlers[TEST_POWER_PATH_NUM_RX_HANDLERS];
+
 void setup_test(void) {
   event_queue_init();
-  gpio_init();
   interrupt_init();
-  gpio_it_init();
   soft_timer_init();
+  gpio_init();
+  gpio_it_init();
   adc_init(ADC_MODE_SINGLE);
+
+  CANSettings can_settings = {
+    .device_id = CAN_DEVICE_CHAOS,
+    .bitrate = CAN_HW_BITRATE_125KBPS,
+    .rx_event = CHAOS_EVENT_CAN_RX,
+    .tx_event = CHAOS_EVENT_CAN_TX,
+    .fault_event = CHAOS_EVENT_CAN_FAULT,
+    .tx = { GPIO_PORT_A, 12 },
+    .rx = { GPIO_PORT_A, 11 },
+    .loopback = true,
+  };
+
+  TEST_ASSERT_OK(
+      can_init(&can_settings, &s_can_storage, s_rx_handlers, SIZEOF_ARRAY(s_rx_handlers)));
   TEST_ASSERT_OK(power_path_init(&s_ppc));
 }
 
 void teardown_test(void) {}
 
 void test_power_path_uv_ov(void) {
+  volatile CANMessage rx_msg = { 0 };
+  can_register_rx_handler(CAN_MESSAGE_OVUV_DCDC_AUX, prv_handle_uvov, &rx_msg);
+
   TEST_ASSERT_OK(power_path_source_monitor_enable(&s_ppc.aux_bat, TEST_POWER_PATH_ADC_PERIOD_US));
   TEST_ASSERT_OK(power_path_source_monitor_enable(&s_ppc.dcdc, TEST_POWER_PATH_ADC_PERIOD_US));
 
+  delay_us(TEST_POWER_PATH_ADC_PERIOD_US + TEST_POWER_PATH_ADC_PERIOD_US / 10);
+
   gpio_it_trigger_interrupt(&s_ppc.aux_bat.uv_ov_pin);
-  Event e = { 0, 0 };
-  while (event_process(&e) != STATUS_CODE_OK) {
-  }
-  TEST_ASSERT_EQUAL(CHAOS_EVENT_CAN_UV_OV, e.id);
-  TEST_ASSERT_EQUAL(POWER_PATH_SOURCE_ID_AUX_BAT, e.data);
+
+  volatile Event e = { 0 };
+  volatile StatusCode status = NUM_STATUS_CODES;
+
+  // TX
+  do {
+    status = event_process(&e);
+  } while (status != STATUS_CODE_OK);
+  TEST_ASSERT_TRUE(fsm_process_event(CAN_FSM, &e));
+  // RX
+  do {
+    status = event_process(&e);
+  } while (status != STATUS_CODE_OK);
+  TEST_ASSERT_TRUE(fsm_process_event(CAN_FSM, &e));
+
+  TEST_ASSERT_EQUAL(false, s_dcdc_uv);
```

Comment:
Maybe I missed it, but how are you ensuring that these faults occur when you trigger an interrupt? Are you assuming a specific hardware configuration for testing?

Reply:
I inject an ADC conversion function that spits out a fixed value. See line 67.